### DISCUSSION TITLE
Chain with a group of a single task

### DISCUSSION THREAD

**Discussion by edwardreed81** [celery/celery#8991] [Q&A]
The workflow I'm constructing is relatively simple:

A setup task that executes, followed by a group of tasks that depend on the result of the setup task:

```python
run_tasks = [run_task(...).s() for task in tasks]
chain = setup_task(...).s() | group(run_tasks)
chain.apply_async()
```

This works fine when `len(run_tasks) > 1`, but when `len(run_tasks) == 1`, constructing the chain changes the second task in the chain to be a regular task, not a group. This means the result of `apply_async()` is a different type depending on the length of `run_tasks`. If it is length 1, the result is an `AsyncResult`, but if it is longer than 1, the result is a `GroupResult`. This causes a problem with saving and restoring the result.

Is this expected behavior? Seems like a group should remain a group no matter how many tasks are in it.

**Comment by mdedonno1337** [comment]
reading the code and the doc, it seems to be expected. apparently, for consistency and simplification of the execution of tasks (for example, a `group() | task` is promoted to a `chord()` automatically.
I suspect there is a way to annotate the extected result with a type more generic.

done in the function https://github.com/celery/celery/blob/8dace6d94b7714db4c677ccea824bc5fee1fdede/celery/canvas.py#L43 I suspect.
